nx/lib/nx/defn/grad.ex

```diff
@@ -424,6 +424,24 @@ defmodule Nx.Defn.Grad do
     to_grad(x, g, cache)
   end
 
+  defp grad(:concatenate, [tensors, axis], %{shape: ans_shape}, g, cache) do
+    zero_axes = List.duplicate(0, tuple_size(ans_shape))
+    ans_shape_list = Tuple.to_list(ans_shape)
+
+    tensors
+    |> Enum.reduce({0, []}, fn t, {limit, result} ->
+      t_len = elem(t.shape, axis)
+      current_limit = t_len + limit
+      start = List.replace_at(zero_axes, axis, limit)
+      len = List.replace_at(ans_shape_list, axis, t_len)
+
+      result = [{t, Nx.slice(g, start, len)} | result]
+      {current_limit, result}
+    end)
+    |> elem(1)
+    |> grad_pairs(g, cache)
```

Comment:
```suggestion
    |> elem(1)
    |> Enum.reverse()
    |> grad_pairs(g, cache)
```

This should be applied if `grad_pairs` depends on the ordering of the tensors. The tests kept passing even without the reverse, so I assumed it wasn't needed.

Reply:
The ordering doesn't matter. grad_pairs will basically call Nx.add/2 after grading each pair.